packages/next/src/export/routes/app-page.ts

```diff
@@ -120,11 +125,39 @@ export async function exportAppPage(
       if (renderOpts.experimental.isRoutePPREnabled) {
         // If PPR is enabled, we should emit the flight data as the prefetch
         // payload.
+        // TODO: This will eventually be replaced by the per-segment prefetch
+        // output below.
         await fileWriter(
           ExportedAppPageFiles.PREFETCH_FLIGHT,
           htmlFilepath.replace(/\.html$/, RSC_PREFETCH_SUFFIX),
           flightData
         )
+
+        if (segmentFlightData) {
+          // Emit the per-segment prefetch data. We emit them as separate files
+          // so that the cache handler has the option to treat each as a
+          // separate entry.
+          segmentPaths = []
+          const segmentsDir = htmlFilepath.replace(
+            /\.html$/,
+            RSC_SEGMENTS_DIR_SUFFIX
+          )
+          const tasks = []
+          for (const [segmentPath, buffer] of segmentFlightData.entries()) {
+            segmentPaths.push(segmentPath)
+            const segmentDataFilePath =
+              segmentPath === '/'
+                ? segmentsDir + '/_index' + RSC_SEGMENT_SUFFIX
+                : segmentsDir + segmentPath + RSC_SEGMENT_SUFFIX
+            tasks.push(
```

Comment:
Are these awaited somewhere to avoid dangling promise?

Reply:
Oh good catch, meant to do `Promise.all` on the line below